File: practice/convert/example2_book_library/java/book-library/src/main/java/dev/starryeye/book_library/application/facade/LoanBookFacade.java
package dev.starryeye.book_library.application.facade;

import dev.starryeye.book_library.application.book.query.GetBookService;
import dev.starryeye.book_library.application.facade.input.LoanBookInput;
import dev.starryeye.book_library.application.user.command.LoanBookService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Transactional
@Component
@RequiredArgsConstructor
public class LoanBookFacade {

    private final GetBookService getBookService;
    private final LoanBookService loanBookService;

    public void loan(LoanBookInput input) {
        if (!getBookService.existsBy(input.bookId())) {
            throw new IllegalArgumentException("book is not found, id = " + input.bookId());
        }

        loanBookService.loan(input.userId(), input.bookId());
    }
}

File: practice/convert/example2_book_library/kotlin/book-library/src/main/kotlin/dev/starryeye/book_library/application/facade/LoanBookFacade.kt
package dev.starryeye.book_library.application.facade

import dev.starryeye.book_library.application.book.query.GetBookService
import dev.starryeye.book_library.application.facade.input.LoanBookInput
import dev.starryeye.book_library.application.user.command.LoanBookService
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Transactional
@Component
class LoanBookFacade(
    private val getBookService: GetBookService,
    private val loanBookService: LoanBookService,
) {

    fun loanBook(input: LoanBookInput) {

        if (!getBookService.existsBy(bookId = input.bookId)) {
            throw IllegalArgumentException("book is not found, id = ${input.bookId}")
        }

        loanBookService.loan(userId = input.userId, bookId = input.bookId)
    }
}